### PR TITLE
Add `TransferPackageController#index` action

### PR DESCRIPTION
**Pull request by straight-shoota**
<del>Builds on top of #1641 and #1625</del>

Rebased on top of #1694

Ref #1612

grafik

### CODE REVIEW

**Comment by straight-shoota** [comment]
Rebased on top of #1694

File: app/presenters/transfer_package_presenter.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+class TransferPackagePresenter
+  attr_reader :transfer_package, :context, :current_user
+
+  def initialize(transfer_package, context)
+    @transfer_package = transfer_package
+    @context = context
+  end
+
+  # TODO(Rails 5.1): Use delegate_missing
+  delegate :uuid, :recipient, :confirmed_at, :confirmed?, :created_at, :receiver_institution, :sender_institution, :sample_transfers, :samples, to: :transfer_package
```
Comment (ysbaddaden) [inline app/presenters/transfer_package_presenter.rb:10]: You can use it today, I backported it in `lib/delegate_missing_to`!
```suggestion
  delegate_missing_to :transfer_package
```

File: app/controllers/transfer_packages_controller.rb [hunk 2]
```diff
@@ -4,6 +4,38 @@ class TransferPackagesController < ApplicationController
...
+    @transfer_packages = TransferPackage
+      .within(@navigation_context.institution)
+      .includes(:receiver_institution, :sender_institution)
+      .order(created_at: :desc)
+
+    if search_uuid = params[:search_uuid].presence
+      @transfer_packages = @transfer_packages.joins(box_transfers: {box: {samples: :sample_identifiers}})
+        .where("transfer_packages.uuid LIKE concat('%', ?, '%') OR boxes.uuid LIKE concat('%', ?, '%') OR sample_identifiers.uuid LIKE concat('%', ?, '%')", search_uuid, search_uuid, search_uuid)
+        .distinct
+    end
```
Comment (ysbaddaden) [inline app/controllers/transfer_packages_controller.rb:17]: Maybe we could simplify this action with scopes on TransferPackage, taking advantage that a scope returns the original query if the lambda returns `nil`:

```ruby
scope :search_uuid, ->(uuid) {
  joins().where().distinct if uuid.present?
}

sample_transfers.search_uuid(params[:search_uuid])
```

File: app/controllers/transfer_packages_controller.rb [hunk 3]
```diff
@@ -4,6 +4,38 @@ class TransferPackagesController < ApplicationController
...
+
+    if institution_query = params[:institution].presence
+      @transfer_packages = @transfer_packages
+        .joins(:receiver_institution, :sender_institution)
+        .where("(institutions.name LIKE concat('%', ?, '%') AND institutions.id != ?) OR (sender_institutions_transfer_packages.name LIKE concat('%', ?, '%') AND sender_institutions_transfer_packages.id != ?)", institution_query, @navigation_context.institution.id, institution_query, @navigation_context.institution.id)
+    end
+
+    @transfer_packages = perform_pagination(@transfer_packages)
+      .preload(:sender_institution, :receiver_institution)
```
Comment (ysbaddaden) [inline app/controllers/transfer_packages_controller.rb:35]: You already `#includes` the association above, so the query will already preload them.
```suggestion
```

**Review by ysbaddaden — APPROVED** [review]

**Comment by ysbaddaden** [comment]
@straight-shoota I found a bug with the `boxes/preview` partial: it generates a link to the box, which is nice, but the sender institution can't access the box anymore (oops).

**Comment by ysbaddaden** [comment]
@straight-shoota well, the box doesn't belong to an institution/site anymore during the transfer, so... neither the sender nor the receiver should access the package contents (neither the boxes nor their samples), until the receiver confirms the reception, in which case the boxes and samples will be accessible to them (the new institution will be set).

Maybe we should remove the link altogether :disappointed_relieved:

**Review by ysbaddaden — APPROVED** [review]

**Pull request merged**